---
title: "DIANE - Differential Expression Analysis report"
output: 
  html_document:
    theme: flatly
    highlight: tango
    code_folding: hide
    df_print: paged
params:
  r_dea: NA
  r: NA
---

```{r, include=FALSE}
r <- params$r_dea
the_r <- params$r
library(knitr)
```

# Dashboard for the Inference and Analysis of Networks from Expression data <img src="favicon.ico" align="right" width="160"/>

------------------------------------------------------------------------

This report was automatically generated by [DIANE](https://oceanecsn.github.io/DIANE) to improve research reproducibility.

It contains the main settings and results for the DEA tab of the application, reporting the last transcriptome comparison that was performed.

# Your settings

Reference and perturbation condition :

```{r}
paste(r$ref, r$trt)
```

Threshold adjusted p-value and minimal expected absolute log fold change :

```{r}
paste("FDR =", r$fdr, "LFC = ", r$lfc)
```

# Results

EdgeR statistical model:

```{r}
r$fit
```

Number of up-regulated genes :

```{r}
nrow(r$gene_table[r$gene_table$Regulation == "Up",])
```

Number of down-regulated genes :

```{r}
nrow(r$gene_table[r$gene_table$Regulation == "Down",])
```

DEGs table :

```{r}
DT::datatable(r$gene_table)
```

The entire table is downloadable from the application as a csv file.

# Plots

```{r, fig.width=10, fig.height=8}
DIANE::draw_DEGs(r$tags, fdr = r$fdr, lfc = r$lfc, MA = TRUE)
DIANE::draw_DEGs(r$tags, fdr = r$fdr, lfc = r$lfc, MA = FALSE)
DIANE::draw_heatmap(r$gene_table$genes, data = the_r$normalized_counts, title = paste(r$ref, "versus", r$trt))
```

## GO enrichment analysis

If you performed GO enrichment analysis, you can directly download the result dataframe from the GO analysis tab in the application.
