---
title: "DIANE - Normalisation report"
output: 
  html_document:
    theme: flatly
    highlight: tango
    code_folding: hide
params:
  r: NA
  input: NA
---

```{r, include=FALSE}
r <- params$r
input <- params$input

library(knitr)
```


# Dashboard for the Inference and Analysis of Networks from Expression data <img src="favicon.ico" align="right" alt="" width="120" />

---

This report was automatically generated by [DIANE](https://oceanecsn.github.io/DIANE) to improve research reproducibility.

It contains the main settings and results for the chosen tab of the application.


# Your data

The organism you're studying :
```{r}
r$organism
```

Number of genes :
```{r}
nrow(r$raw_counts)
```

Experimental conditions :
```{r}
unique(r$conditions)
```

Design:
```{r}
r$design
```



# Settings


Prior removal of differentially expressed genes :
```{r}
input$prior_removal
```

Normalisation method :
```{r}
input$norm_method
```

Number of genes before low counts removal :
```{r}
nrow(r$raw_counts)
```

Number of genes after low counts removal :
```{r}
nrow(r$normalized_counts)
```

The criteria for filtering was a sum of expression accross all samples above 
```{r}
input$low_counts_filter
```

## Distribution visualisation

For each condition, we can visualize the distributions of gene counts before normalisation and low can removal, and after.

```{r viz_filter, fig.width=10, fig.height=12}
pre_process <- DIANE::draw_distributions(r$raw_counts, boxplot = FALSE) + ggplot2::ggtitle("Before")
post_process <- DIANE::draw_distributions(r$normalized_counts, boxplot = FALSE)+ ggplot2::ggtitle("After")

gridExtra::grid.arrange(pre_process, post_process, ncol = 2)

```

The distribution modes should be aligned, and the initial peak around low values should have disappeared.

## Summary

Raw TCC summary including the chosen pipeline, library sizes and normalization factors :

```{r}
print(r$tcc)
```


# Exploratory analysis

Once normalisation is complete, one can start by exploraing in a global way which are the experimental conditions that drive gene expression. It is typically conducted by dimensionality reduction strategies as follows :

## PCA


```{r, fig.width=10, fig.height=8}
draw_PCA(r$normalized_counts)
```


## MDS


```{r, fig.width=10, fig.height=8}
draw_MDS(r$normalized_counts)
```